src/dmqproto/client/request/internal/Consume.d

```diff
@@ -12,6 +12,711 @@
+
+            Notification notification;
+            notification.received = RequestDataInfo(this.outer.context.request_id, record);
+            Consume.notify(this.outer.context.user_params, notification);
+
+            if (!initially_stopped &&
+                this.outer.context.shared_working.suspendable_control.stopped
+            )
+                this.outer.sendStopMessage(this.fiber);
+        }
+    }
+
+    /***************************************************************************
+
+        The fiber that reads messages from the node and notifies `RecordStream`.
+
+    ***************************************************************************/
+
+    private class Reader
+    {
+        /// The fiber.
+        private MessageFiber fiber;
+
+        /// The `RecordStream` to notify when a message has arrived.
+        private RecordStream record_stream;
+
+        /***********************************************************************
+
+            Constructor, starts the fiber.
+
+            Params:
+                record_stream = the `RecordStream` to notify when a message has
+                    arrived
+
+        ***********************************************************************/
+
+        private this ( RecordStream record_stream )
+        {
+            this.record_stream = record_stream;
+            this.fiber = this.outer.resources.getFiber(&this.fiberMethod);
+            this.fiber.start();
+        }
+
+        /***********************************************************************
+
+            Reads messages from the node and notifies `record_strem` by calling
+            its respective methods when a `Records` or `Stopped` message has
+            arrived. Quits when a `Stopped` message has arrived because
+            `Stopped` is the last message from the node.
+
+        ***********************************************************************/
+
+        private void fiberMethod ( )
+        {
+            while (true)
+            {
+                auto msg = this.outer.request_event_dispatcher.receive(
+                    this.fiber,
+                    Message(MessageType_v3.Records),
+                    Message(MessageType_v3.Stopped),
+                    Message(MessageType_v3.ChannelRemoved)
+                );
+
+                switch (msg.type)
+                {
+                    case MessageType_v3.Records:
+                        Const!(void)[] received_record_batch;
+                        this.outer.conn.message_parser.parseBody(
+                            msg.payload, received_record_batch
+                        );
+                        this.record_stream.addRecords(received_record_batch);
+                        break;
+
+                    case MessageType_v3.Stopped:
+                        this.record_stream.stop();
+                        return;
+
+                    case MessageType_v3.ChannelRemoved:
+                        // TODO
```

Comment:
Thinking about ChannelRemoved handling at a high level, it should probably cause the request to restart (resending the initial payload), essentially, which will cause the channel to be recreated. I guess when this happens, though, we need to make sure any records already received and waiting to be passed to the user are either stored or processed, before the request is restarted. :thinking:

Reply:
Having the batch stored in a shared resource probably means this is possible, though we'd need to make sure the same resource is reused after restarting the request.